extras/src/main/scala/spire/math/extras/interval/IntervalSeq.scala

```diff
@@ -0,0 +1,687 @@
+package spire.math.extras.interval
+
+import java.util.Arrays
+
+import spire.algebra.{Bool, Eq, Order}
+import spire.math.interval._
+import spire.math.{Interval, Rational}
+
+import scala.annotation.tailrec
+import scala.language.implicitConversions
+import scala.reflect.ClassTag
+
+final class IntervalSeq[T] private (
+    val belowAll: Boolean,
+    private val values: Array[T],
+    private val kinds: Array[Byte],
+    private implicit val order: Order[T]) extends IntervalSet[T, IntervalSeq[T]] { lhs =>
+
+  import IntervalSeq._
+
+  private def binarySearch(key: T): Int = {
+    var low = 0
+    var high = values.length - 1
+    while (low <= high) {
+      val mid = (low + high) >>> 1
+      val midVal = values(mid)
+      val c = order.compare(midVal, key)
+      if (c < 0) {
+        low = mid + 1
+      }
+      else if (c > 0) {
+        high = mid - 1
+      }
+      else {
+        // scalastyle:off return
+        return mid
+        // scalastyle:on return
+      }
+    }
+    -(low + 1)
+  }
+
+  def below(index: Int): Boolean = {
+    if (index == 0)
+      belowAll
+    else
+      valueAbove(kinds(index - 1))
+  }
+
+  def at(value: T): Boolean = {
+    val index = binarySearch(value)
+    if (index >= 0)
+      valueAt(kinds(index))
+    else
+      below(-index - 1)
+  }
+
+  def above(value: T): Boolean = {
+    val index = binarySearch(value)
+    if (index >= 0)
+      valueAbove(kinds(index))
+    else
+      below(-index - 1)
+  }
+
+  def below(value: T): Boolean = {
+    val index = binarySearch(value)
+    if (index > 0)
+      valueAbove(kinds(index - 1))
+    else if (index == 0)
+      belowAll
+    else
+      below(-index - 1)
+  }
+
+  def apply(value:T) = at(value)
```

Comment:
We should probably add return types to all the public methods here. Most have them, but a few are missing.